---
title: Kusto. ошибки приема & исключения в Azure обозреватель данных
description: В этой статье описывается Kusto. прием-ошибок и исключений в Azure обозреватель данных.
services: data-explorer
author: orspod
ms.author: orspodek
ms.reviewer: rkarlin
ms.service: data-explorer
ms.topic: reference
ms.date: 10/30/2019
ms.openlocfilehash: 61c183f11aa7658faba00c5dd3c4795f235e5467
ms.sourcegitcommit: 898f67b83ae8cf55e93ce172a6fd3473b7c1c094
ms.translationtype: MT
ms.contentlocale: ru-RU
ms.lasthandoff: 10/21/2020
ms.locfileid: "92337488"
---
# <a name="kustoingest-errors-and-exceptions"></a>Ошибки и исключения Kusto. приема
Любая ошибка во время обработки приема на стороне клиента обозначается исключением C#.

## <a name="failures"></a>Сбои

### <a name="kustodirectingestclient-exceptions"></a>Исключения Кустодиректинжестклиент

При попытке приема из нескольких источников в процессе приема могут возникать ошибки. Если прием для одного из источников завершается неудачей, он заносится в журнал, а клиент продолжит прием оставшихся источников. После перебора всех источников для приема `IngestClientAggregateException` создается исключение, содержащее `IList<IngestClientException> IngestionErrors` член.

`IngestClientException` и его производные классы содержат поле `IngestionSource` и `Error` поле. Эти два поля вместе создают сопоставление с ошибкой, возникшей при попытке приема, из источника, который не удалось принять. Эти сведения можно использовать в списке для выяснения того `IngestionErrors` , какие источники не прошли прием и почему. `IngestClientAggregateException`Исключение также содержит логическое свойство `GlobalError` , которое указывает, произошла ли ошибка для всех источников.

### <a name="failures-ingesting-from-files-or-blobs"></a>Сбои при получении из файлов или больших двоичных объектов

Если при попытке приема из большого двоичного объекта или файла происходит сбой приема, источники приема не удаляются, даже если `deleteSourceOnSuccess` для флага задано значение `true` . Источники сохраняются для дальнейшего анализа. После того как происхождение ошибки будет понята, и если ошибка не возникла в источнике приема, то пользователь клиента может попытаться его повторного приема.

### <a name="failures-ingesting-from-idatareader"></a>Сбои при получении из IDataReader

При приеме из DataReader данные для приема сохраняются во временной папке, расположение которой по умолчанию — `<Temp Path>\Ingestions_<current date and time>` . Эта папка по умолчанию всегда удаляется после успешного приема.

В `IngestFromDataReader` `IngestFromDataReaderAsync` методах и `retainCsvOnFailure` флаг, значение по умолчанию которого равно `false` , определяет, следует ли сохранять файлы после неудачного приема. Если этот флаг установлен в значение `false` , данные, которые не удается принять, не будут сохранены, что затрудняет понимание того, что пошло не так.

## <a name="kustoqueuedingestclient-exceptions"></a>Исключения Кустокуеуединжестклиент

`KustoQueuedIngestClient` принимает данные путем отправки сообщений в очередь Azure. Если ошибка возникает до или во время процесса очереди, `IngestClientAggregateException` в конце процесса создается исключение. Созданное исключение включает коллекцию `IngestClientException` , которая содержит источник каждой ошибки и не была отправлена в очередь. Также возникает ошибка, которая возникла при попытке опубликовать сообщение.

### <a name="posting-to-queue-failures-with-a-file-or-blob-as-a-source"></a>Отправка сообщений о сбоях в очереди с помощью файла или большого двоичного объекта в качестве источника

Если при использовании методов возникла ошибка `KustoQueuedIngestClient` `IngestFromFile/IngestFromBlob` , источники не удаляются, даже если `deleteSourceOnSuccess` для флага задано значение `true` . Вместо этого источники сохраняются для дальнейшего анализа. 

После того как происхождение ошибки будет понята и ошибка не поступила из самого источника приема, пользователь клиента может попытаться повторно поставить данные в очередь, используя соответствующие `IngestFromFile/IngestFromBlob` методы с источником, для которого произошел сбой. 

### <a name="posting-to-queue-failures-with-idatareader-as-a-source"></a>Отправка сообщений о сбоях очереди с помощью IDataReader в качестве источника

При использовании источника DataReader данные, которые необходимо поместить в очередь, сохраняются во временной папке, расположение по умолчанию которой — `<Temp Path>\Ingestions_<current date and time>` . Эта папка всегда удаляется после успешной отправки данных в очередь.
В `IngestFromDataReader` `IngestFromDataReaderAsync` методах и `retainCsvOnFailure` флаг, значение по умолчанию которого равно `false` , определяет, следует ли сохранять файлы после неудачного приема. Если этот флаг установлен в значение `false` , данные, которые не удается принять, не будут сохранены, что затрудняет понимание того, что пошло не так.

### <a name="common-failures"></a>Распространенные сбои
|Error                         |Причина           |Меры по снижению риска                                   |
|------------------------------|-----------------|---------------------------------------------|
|Имя базы данных <database name> не существует| База данных не существует|Проверьте имя базы данных в `kustoIngestionProperties` /CREATE |
|Сущность "имя таблицы, не существующая" типа "Table", не найдена.|Таблица не существует, сопоставление CSV отсутствует.| Добавить сопоставление CSV-файла или создать необходимую таблицу |
|BLOB-объект <blob path> , исключенный по причине: шаблон JSON должен быть принят с помощью параметра жсонмаппинг| Прием JSON, если сопоставление JSON не предоставлено.|Предоставление сопоставления JSON |
|Не удалось скачать BLOB-объект: "удаленный сервер вернул ошибку: (404) не найден".| большой двоичный объект не существует;|Убедитесь, что большой двоичный объект существует. Если он существует, повторите попытку и обратитесь к команде Kusto |
|Сопоставление столбцов JSON недопустимо: два или более элемента сопоставления указывают на один и тот же столбец.| Сопоставление JSON содержит 2 столбца с разными путями|Исправить сопоставление JSON |
|Енгиниррор-[Утилсексцептион] `IngestionDownloader.Download` : не удалось скачать один или несколько файлов (Поиск кустологс для ActivityId: <GUID1> , RootActivityId: <GUID2> )| Не удалось скачать один или несколько файлов. |Повторить попытку |
|Не удалось выполнить синтаксический анализ: поток с ИДЕНТИФИКАТОРом " <stream name> " имеет неправильный формат CSV, сбой на политику валидатионоптионс |Неправильно сформированный CSV-файл (например, не имеет одинакового количества столбцов в каждой строке). Завершается сбоем, только если для политики проверки задано значение `ValidationOptions` . валидатексвинпутконстантколумнс |Проверьте CSV-файлы. Это сообщение относится только к файлам CSV/TSV |
|`IngestClientAggregateException` с сообщением об ошибке "отсутствуют обязательные параметры для допустимой подписи общего доступа" |Используемая SAS относится к службе, а не к учетной записи хранения. |Использование SAS учетной записи хранения |

### <a name="ingestion-error-codes"></a>Коды ошибок приема

Чтобы упростить обработку сбоев приема программными средствами, сведения об ошибках дополнены числовым кодом ошибки ( `IngestionErrorCode enumeration` ).

|ErrorCode                                      |Причина                                                        |
|-----------------------------------------------|--------------------------------------------------------------|
|Неизвестно                                        | Произошла неизвестная ошибка|
|Stream_LowMemoryCondition                      | Операция исчерпала память|
|Stream_WrongNumberOfFields                     | В документе CSV имеется несоответствие количества полей|
|Stream_InputStreamTooLarge                     | Документ, отправленный для приема, превысил допустимый размер|
|Stream_NoDataToIngest                          | Не найдены потоки данных для приема|
|Stream_DynamicPropertyBagTooLarge              | Один из динамических столбцов в полученных данных содержит слишком много уникальных свойств|
|Download_SourceNotFound                        | Не удалось скачать источник из службы хранилища Azure-источник не найден|
|Download_AccessConditionNotSatisfied           | Не удалось скачать источник из службы хранилища Azure — доступ запрещен|
|Download_Forbidden                             | Не удалось скачать источник из службы хранилища Azure — доступ запрещен|
|Download_AccountNotFound                       | Не удалось скачать источник из хранилища Azure — учетная запись не найдена|
|Download_BadRequest                            | Не удалось скачать источник из службы хранилища Azure — неправильный запрос|
|Download_NotTransient                          | Не удалось скачать источник из службы хранилища Azure — невременная ошибка|
|Download_UnknownError                          | Не удалось скачать источник из службы хранилища Azure — неизвестная ошибка|
|UpdatePolicy_QuerySchemaDoesNotMatchTableSchema| Не удалось вызвать политику обновления. Схема запроса не соответствует схеме таблицы|
|UpdatePolicy_FailedDescendantTransaction       | Не удалось вызвать политику обновления. Не удалось изменить политику обновления потомков|
|UpdatePolicy_IngestionError                    | Не удалось вызвать политику обновления. Произошла ошибка приема|
|UpdatePolicy_UnknownError                      | Не удалось вызвать политику обновления. Произошла неизвестная ошибка|
|BadRequest_MissingJsonMappingtFailure          | Шаблон JSON не получил параметр Жсонмаппинг|
|BadRequest_InvalidBlob                         | Подсистеме не удалось открыть и прочитать не ZIP-файл BLOB|
|BadRequest_EmptyBlob                           | Пустой BLOB-объект|
|BadRequest_EmptyArchive                        | ZIP-файл не содержит заархивированных элементов|
|BadRequest_EmptyBlobUri                        | Указанный URI большого двоичного объекта пуст|
|BadRequest_DatabaseNotExist                    | База данных не существует|
|BadRequest_TableNotExist                       | Таблица не существует|
|BadRequest_InvalidKustoIdentityToken           | Недопустимый токен удостоверения Kusto|
|BadRequest_UriMissingSas                       | Путь к BLOB-объекту без SAS из неизвестного хранилища BLOB-объектов|
|BadRequest_FileTooLarge                        | Попытка приема слишком большого файла|
|BadRequest_NoValidResponseFromEngine           | Нет допустимого ответа от принимающей команды|
|BadRequest_TableAccessDenied                   | Отказано в доступе к таблице|
|BadRequest_MessageExhausted                    | Сообщение исчерпано|
|General_BadRequest                             | Общий недопустимый запрос. Может подсказка о приеме несуществующей базы данных или таблицы|
|General_InternalServerError                    | Произошла внутренняя ошибка сервера|

## <a name="detailed-exceptions-reference"></a>Подробные справочные сведения об исключениях

### <a name="cloudqueuesnotfoundexception"></a>клаудкуеуеснотфаундексцептион

Возникает, если из кластера Управление данными не были возвращены очереди

Базовый класс: [исключение](/dotnet/api/system.exception)

|Имя поля |Тип     |Значение
|-----------|---------|------------------------------|
|Error      | Строка  | Ошибка, возникшая при попытке получить очереди из DM
                            
Имеет смысл только при использовании [клиента, принимающего очереди Kusto](kusto-ingest-client-reference.md#interface-ikustoqueuedingestclient).
Во время приема для получения очередей Azure, связанных с DM, выполняются несколько попыток. При сбое этих попыток в поле "ошибка" возникает исключение, содержащее причину сбоя. Возможно, также вызывается внутреннее исключение в поле "InnerException".


### <a name="cloudblobcontainersnotfoundexception"></a>клаудблобконтаинерснотфаундексцептион

Возникает, когда из кластера Управление данными не были возвращены контейнеры больших двоичных объектов

Базовый класс: [исключение](/dotnet/api/system.exception)

|Имя поля   |Тип     |Значение       
|-------------|---------|------------------------------|
|кустоендпоинт| Строка  | Конечная точка соответствующего DM
                            
Имеет смысл только при использовании [клиента, принимающего очереди Kusto](kusto-ingest-client-reference.md#interface-ikustoqueuedingestclient).  
При приеме источников, которые еще не находятся в контейнере Azure, например в файлах, DataReader или Stream, данные отправляются во временный большой двоичный объект для приема. Исключение возникает, если не найдены контейнеры для передачи данных.

### <a name="duplicateingestionpropertyexception"></a>дупликатеинжестионпропертексцептион

Возникает, если свойство приема настроено более одного раза

Базовый класс: [исключение](/dotnet/api/system.exception)

|Имя поля   |Тип     |Значение       
|-------------|---------|------------------------------------|
|PropertyName | Строка  | Имя повторяющегося свойства
                            
### <a name="postmessagetoqueuefailedexception"></a>постмессажетокуеуефаиледексцептион

Возникает при сбое отправки сообщения в очередь

Базовый класс: [исключение](/dotnet/api/system.exception)

|Имя поля   |Тип     |Значение       
|-------------|---------|---------------------------------|
|куеуеури     | Строка  | URI очереди
|Error        | Строка  | Сообщение об ошибке, созданное при попытке публикации в очереди
                            
Имеет смысл только при использовании [клиента, принимающего очереди Kusto](kusto-ingest-client-reference.md#interface-ikustoqueuedingestclient).  
Принимающий постановкой клиент принимает данные путем отправки сообщения в соответствующую очередь Azure. При возникновении ошибки POST возникает исключение. Он будет содержать URI очереди, причину сбоя в поле "ошибка" и, возможно, внутреннее исключение в поле "InnerException".

### <a name="dataformatnotspecifiedexception"></a>датаформатнотспеЦифиедексцептион

Возникает, если формат данных обязателен, но не указан в `IngestionProperties`

Базовый класс: Инжестклиентексцептион

При приеме из потока формат данных должен быть указан в [инжестионпропертиес](kusto-ingest-client-reference.md#class-kustoingestionproperties)для правильного приема данных. Это исключение возникает, если `IngestionProperties.Format` не указано.

### <a name="invaliduriingestclientexception"></a>инвалидуриинжестклиентексцептион

Возникает при отправке недопустимого URI BLOB-объекта в качестве источника приема

Базовый класс: Инжестклиентексцептион

### <a name="compressfileingestclientexception"></a>компрессфилеинжестклиентексцептион

Возникает, когда принимающий клиент не может сжать файл, предоставленный для приема

Базовый класс: Инжестклиентексцептион

Файлы сжимаются до их приема. Исключение возникает, когда попытка сжатия файла завершается ошибкой.

### <a name="uploadfiletotempblobingestclientexception"></a>уплоадфилетотемпблобинжестклиентексцептион

Возникает, когда принимающий клиент не может отправить источник, предоставленный для приема во временный большой двоичный объект

Базовый класс: Инжестклиентексцептион

### <a name="sizelimitexceededingestclientexception"></a>сизелимитексцеедединжестклиентексцептион

Возникает, когда источник приема слишком велик

Базовый класс: Инжестклиентексцептион

|Имя поля   |Тип     |Значение       
|-------------|---------|-----------------------|
|Размер         | long    | Размер источника приема
|MaxSize      | long    | Максимальный размер, допустимый для приема

Если источник приема превышает максимальный размер, равный 4 ГБ, возникает исключение. Проверка размера может быть переопределена `IgnoreSizeLimit` флагом в [классе инжестионпропертиес](kusto-ingest-client-reference.md#class-kustoingestionproperties). Однако не рекомендуется принимать [отдельные источники, размер которых превышает 1 ГБ](about-kusto-ingest.md#ingestion-best-practices).

### <a name="uploadfiletotempblobingestclientexception"></a>уплоадфилетотемпблобинжестклиентексцептион

Возникает, когда принимающий клиент не может отправить файл, предоставленный для приема во временный большой двоичный объект

Базовый класс: Инжестклиентексцептион

### <a name="directingestclientexception"></a>директинжестклиентексцептион

Возникает при возникновении общей ошибки во время прямого приема

Базовый класс: Инжестклиентексцептион

### <a name="queuedingestclientexception"></a>куеуединжестклиентексцептион

Возникает при возникновении ошибки во время приема в очереди

Базовый класс: Инжестклиентексцептион

### <a name="ingestclientaggregateexception"></a>инжестклиентаггрегатиксцептион

Возникает при возникновении одной или нескольких ошибок во время приема

Базовый класс: [AggregateException](/dotnet/api/system.aggregateexception)

|Имя поля      |Тип                             |Значение       
|----------------|---------------------------------|-----------------------|
|инжестионеррорс | IList<IngestClientException>    | Ошибки, возникающие при попытке приема, и источники, связанные с ними
|исглобалеррор   | bool                            | Указывает, произошло ли исключение для всех источников

## <a name="next-steps"></a>Next Steps

Дополнительные сведения об ошибках в машинном коде см. [в разделе ошибки в машинном коде](../../concepts/errorsinnativecode.md).